src/blockstore/database.zig

```diff
@@ -0,0 +1,248 @@
+            return try self.impl.put(cf, comptime cf.find(column_families), key, value);
+        }
+
+        pub fn get(self: *Self, comptime cf: ColumnFamily, key: cf.Key) !?cf.Value {
+            return try self.impl.get(cf, comptime cf.find(column_families), key);
+        }
+
+        /// Returns a reference to the serialized bytes.
+        ///
+        /// This is useful in two situations:
+        ///
+        /// 1. You don't plan to deserialize the data, and just need the bytes.
+        ///
+        /// 2. `cf.Value` is []const u8, and you don't need an owned slice. In this
+        ///    case, getBytes is faster than get. But if you *do* need an owned slice,
+        ///    then it's faster to call `get` insted of calling this function followed
+        ///    by memcpy.
+        pub fn getBytes(
+            self: *Self,
+            comptime cf: ColumnFamily,
+            key: cf.Key,
+        ) !?BytesRef {
+            return try self.impl.getBytes(cf, comptime cf.find(column_families), key);
+        }
+
+        pub fn delete(self: *Self, comptime cf: ColumnFamily, key: cf.Key) !void {
+            return try self.impl.delete(cf, comptime cf.find(column_families), key);
+        }
+
+        pub fn initBatch(self: *Self) !Batch {
+            return .{ .impl = self.impl.initBatch() };
+        }
+
+        pub fn commit(self: *Self, batch: Batch) !void {
+            return self.impl.commit(batch.impl);
+        }
+
+        pub fn runTest() !void {
+            const Value = struct { hello: u16 };
+            const cf1 = ColumnFamily{
+                .name = "one",
+                .Key = u64,
+                .Value = Value,
+            };
+            const cf2 = ColumnFamily{
+                .name = "two",
+                .Key = u64,
+                .Value = Value,
+            };
+            const allocator = std.testing.allocator;
+            const logger = Logger.init(std.testing.allocator, Logger.TEST_DEFAULT_LEVEL);
+            defer logger.deinit();
+            var db = try Database(Impl, &.{ cf1, cf2 }).open(
+                allocator,
+                logger,
+                "test_data/bsdb",
+            );
+            defer db.deinit();
+            try db.put(cf1, 123, .{ .hello = 345 });
+            const got = try db.get(cf1, 123);
+            try std.testing.expect(345 == got.?.hello);
+            const not = try db.get(cf2, 123);
+            try std.testing.expect(null == not);
+            const wrong_was_deleted = try db.delete(cf2, 123);
+            _ = wrong_was_deleted;
+            // try std.testing.expect(!wrong_was_deleted); // FIXME
+            const was_deleted = try db.delete(cf1, 123);
+            _ = was_deleted;
+            // try std.testing.expect(was_deleted);
+            const not_now = try db.get(cf1, 123);
+            try std.testing.expect(null == not_now);
+        }
+    };
+}
+
+pub fn BatchImpl(
```

Comment:
Regarding the separation between Batch and Database, they inevitably need to be two separate structs, regardless of whether the abstraction layer exists. In rocksdb itself, a write batch is a distinct struct with separate put and delete methods from the database.

Once I start working on the performance of the blockstore, I was planning to swap rocksdb out with alternatives. lmdb seems like it could be better. I don't think this is very far away. I'm not sure what you mean about it being high risk. I do see vendor lock-in as a risk though.

I understand that abstractions like this have a cost of additional click-throughs. I guess I just see it as a small cost in exchange for a large benefit.

That said, I have some vague ideas about a way to refactor this that might address your concerns, without abandoning the abstraction. Basically it would mean pushing the abstraction to the side, instead of using it as a middleman that obscures control flow. Initially it didn't seem possible, but maybe it's doable. I'll mess with the code a bit and see if I can make it work.

Reply:
I have refactored the code to address what I believe are your primary concerns:
1. Inconsistent naming and excessive type aliasing makes it confusing to use write batches
2. It would be easier to follow the blockstore control flow if the blockstore could directly use RocksDB instead of needing to go through an abstract Database struct as an intermediary.

1  was straightforward. I moved the batch struct definitions into the database struct definitions, instead of using a type alias. Also I consolidated everything onto the name "WriteBatch" so there is no more `Batch` `BatchImpl` etc.

For 2, I'm actually not sure if I've properly understood your concern. But anyway I came up with an approach here so that the RocksDB struct can be directly used as the database, without needing to declare `Database` as the actual database implementation. The key change is that we have this code now:

https://github.com/Syndica/sig/blob/9be5d821a6487c20131c09f2bc8854d732f5ccfd/src/blockstore/blockstore.zig#L3

instead of this:

```zig
pub fn BlockstoreDB(comptime DB: type) type {
    return sig.blockstore.database
        .Database(DB, &sig.blockstore.schema.list);
}
```

Database itself still exists, and can be used as a RocksDB wrapper. But now it's only used for tests. 

I'm actually not convinced this approach is better. I prefer to depend on abstractions rather than concrete implementations in cases like this. But I'm OK with it because I can see the advantage, and I found a way to retain the most valuable aspects of the abstraction.

Basically the change I made was to ensure that Database, RocksDB, and SharedHashMapDB all have truly identical interfaces. Also I made it so each database impl stands alone, as in, I stripped out whatever minimal logic was in Database, and moved it to the impls (such as finding the cf index). That way you can actually swap out each of these structs for one another. Any of them could work as the blockstore backend.

I also implemented some validations to guarantee that the structs are (and will remain) fully interchangeable. You can see that code here:
https://github.com/Syndica/sig/blob/9be5d821a6487c20131c09f2bc8854d732f5ccfd/src/utils/interface.zig#L3

It's used here. And this function is called when running the database tests for a particular implementation.
https://github.com/Syndica/sig/blob/9be5d821a6487c20131c09f2bc8854d732f5ccfd/src/blockstore/database.zig#L8-L12